machine_translation/sampling.py

```diff
@@ -0,0 +1,307 @@
+
+        if self.config['reload']:
+            try:
+                bleu_score = numpy.load(os.path.join(self.config['saveto'],
+                                        'val_bleu_scores.npz'))
+                self.val_bleu_curve = bleu_score['bleu_scores'].tolist()
+
+                # Track n best previous bleu scores
+                for i, bleu in enumerate(
+                        sorted(self.val_bleu_curve, reverse=True)):
+                    if i < self.track_n_models:
+                        self.best_models.append(ModelInfo(bleu))
+                logger.info("BleuScores Reloaded")
+            except:
+                logger.info("BleuScores not Found")
+
+    def do(self, which_callback, *args):
+
+        # Track validation burn in
+        if self.main_loop.status['iterations_done'] <= \
+                self.config['val_burn_in']:
+            return
+
+        # Get current model parameters
+        self.model.set_param_values(
+            self.main_loop.model.get_param_values())
+
+        # Evaluate and save if necessary
+        self._save_model(self._evaluate_model())
+
+    def _evaluate_model(self):
+
+        logger.info("Started Validation: ")
+        val_start_time = time.time()
+        mb_subprocess = Popen(self.multibleu_cmd, stdin=PIPE, stdout=PIPE)
+        total_cost = 0.0
+
+        # Get target vocabulary
+        if not self.trg_ivocab:
+            sources = self._get_attr_rec(self.main_loop, 'data_stream')
+            trg_vocab = sources.data_streams[1].dataset.dictionary
+            self.trg_ivocab = {v: k for k, v in trg_vocab.items()}
+
+        if self.verbose:
+            ftrans = open(self.config['val_set_out'], 'w')
+
+        for i, line in enumerate(self.data_stream.get_epoch_iterator()):
+            """
+            Load the sentence, retrieve the sample, write to file
+            """
+
+            seq = self._oov_to_unk(line[0])
+            input_ = numpy.tile(seq, (self.config['beam_size'], 1))
+
+            # draw sample, checking to ensure we don't get an empty string back
+            trans, costs = \
+                self.beam_search.search(
+                    input_values={self.source_sentence: input_},
+                    max_length=3*len(seq), eol_symbol=self.eos_idx,
+                    ignore_first_eol=True)
+
+            nbest_idx = numpy.argsort(costs)[:self.n_best]
+            for j, best in enumerate(nbest_idx):
+                try:
+                    total_cost += costs[best]
+                    trans_out = trans[best]
+
+                    # convert idx to words
+                    trans_out = self._idx_to_word(trans_out, self.trg_ivocab)
+
+                except ValueError:
+                    print "Can NOT find a translation for line: {}".format(i+1)
+                    trans_out = '<UNK>'
+
+                if j == 0:
+                    # Write to subprocess and file if it exists
+                    print >> mb_subprocess.stdin, trans_out
```

Comment:
We use python3-style print everywhere else (`from __future__ import print_function`)